File: 515. Find Largest Value in Each Tree Row.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> largestValues(TreeNode root) {
        if(root == null) return Collections.emptyList();
        
        List<Integer> result = new LinkedList<>();
        Queue<TreeNode> q = new LinkedList<>();
        
        q.offer(root);

        while(!q.isEmpty()){
            int qSize = q.size();
            int rowMax = Integer.MIN_VALUE;

            while(qSize-- > 0){
                TreeNode item = q.poll();
                rowMax = Math.max(rowMax, item.val);
                if(item.left != null) q.offer(item.left);
                if(item.right != null) q.offer(item.right);
            }

            result.add(rowMax);
        }

        return result;
        
    }
}

File: 515. Find Largest Value in Each Tree Row.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    /*
    Time: O(n)
    Space: O(h) (h = height of the tree)

    Do BFS level by level, and track maximum for each level

    */
    vector<int> largestValues(TreeNode* root) {
        vector<int> ret;    
        if(root == nullptr)  return {};

        //BFS
        int level = 0;
        queue<TreeNode*> q;
        q.push(root);
        
        while(!q.empty()){
            queue<TreeNode*> tmp;
           
            while(!q.empty()){
                auto node = q.front();
                q.pop();
                
                if(ret.size()==level)  ret.push_back(node->val);
                else  ret[level] = max(ret[level], node->val);

                if(node->left != nullptr)  tmp.push(node->left);
                if(node->right != nullptr)  tmp.push(node->right);
                
            }

            level++;
            q = move(tmp);
        }

        return ret;
    }
};
